### PR TITLE
feat: Add MySQL instance address attribute

### PR DESCRIPTION
**Pull request by arielvalentin**
Fixes # N/A

## Changes

At GitHub we run MySQL in multi-node clusters and are often interested in knowing the hostname of a node that is executing a query or command.

In order to acheive this we added an attribute to the Trilogy instrumentation named `db.mysql.instance.address` which is populated with the hostname of the node:

<https://github.com/open-telemetry/opentelemetry-ruby-contrib/pull/487>

This change adds the attribute to the database model as a MySQL specific connection attribute.

## Merge requirement checklist

* [x] [CONTRIBUTING.md](https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md) guidelines followed.
* [ ] [CHANGELOG.md](https://github.com/open-telemetry/semantic-conventions/blob/main/CHANGELOG.md) updated for non-trivial changes.
* [ ] [schema-next.yaml](https://github.com/open-telemetry/semantic-conventions/blob/main/schema-next.yaml) updated with changes to existing conventions.

### CODE REVIEW

File: docs/database/mysql.md
Comment (pyohannes) [inline docs/database/mysql.md:21]: Do you, given the proposals in #321 (e. g. https://github.com/open-telemetry/semantic-conventions/pull/321#issuecomment-1720609696), that `server.address` could be sufficient to model this?

File: docs/database/mysql.md
Comment (pyohannes) [inline docs/database/mysql.md:9]:
```suggestion
The Semantic Conventions for the [MySQL](https://www.mysql.com/) extend the [Database Semantic Conventions](database-spans.md)
```

A nit, but I think product-specific semantic conventions shouldn't _override_ general domain-specific ones. General domain-specific semantic conventions should apply to all related products in the domain and allow consumers to interpret data without considering any product-specific details.

File: docs/database/mysql.md
Comment (arielvalentin) [reply docs/database/mysql.md:21]: Some other options I can think of for a more general attribute name:

- `db.instance.server.address`
- `db.node.server.address`

Comment (pyohannes) [reply docs/database/mysql.md:21]: Got it, so `db.mysql.instance.address` will identify the node inside the MySQL cluster. And `server.address` will be the address the user connects to.

Comment (arielvalentin) [reply docs/database/mysql.md:21]: Yes!

**Comment by arielvalentin** [comment]
@pyohannes @AlexanderWert Is there something more that you need from me? Are you wanting to explore the route of generic attribute naming?

**Comment by Oberon00** [comment]
As I understand the new address conventions

> The hostname of the MySQL server instance bound to the current connection. This is useful in cases where the client is connecting to a MySQL server via a proxy or in a clustered environment

This should be filled in in server.address. The current definition of server.address might need to be / have already implicitly changed. See #342.

CC @trask @lmolkova It seems to me that the new address conventions should make this PR redundant, maybe this request could inform / validate the design of #342.

**Comment by pyohannes** [comment]
@Oberon00 

> This should be filled in in server.address. The current definition of server.address might need to be / have already implicitly changed. See https://github.com/open-telemetry/semantic-conventions/pull/342.

I had the same thought, but there's more to it: https://github.com/open-telemetry/semantic-conventions/pull/345#discussion_r1337154428:

> Got it, so `db.mysql.instance.address` will identify the node inside the MySQL cluster. And `server.address` will be the address the user connects to.

File: docs/database/mysql.md
Comment (arielvalentin) [reply docs/database/mysql.md:21]: Naming is notoriously difficult so what do you all think of one of the following:

- `db.instance.address`
- `db.instance.server.address`
- `db.node.address`
- `db.node.server.address`
- `db.server.address`

Comment (pyohannes) [reply docs/database/mysql.md:21]: Mind that I'm not much of a database expert, but `db.instance.address` makes the most sense to me.

**Comment by arielvalentin** [comment]
cc: @cschleiden @andyedison @adrianna-chang-shopify @open-telemetry/ruby-contrib-approvers

File: model/trace/database.yaml
Comment (AlexanderWert) [inline model/trace/database.yaml:252]: If we want to use that attribute in a generic way to identify an instance within a DB cluster, IMHO `db.instance.address` is not the best name, as it's too prescriptive of requiring an address.

For example we cannot use it to replace `db.mssql.instance_name` or `db.elasticsearch.node.name` as those are not necessarily adresses.

How about something more generic that would allow to put either an address, and id or name?

Something like `db.instance.id` ?

**Review by AlexanderWert — CHANGES_REQUESTED** [review]
See my inline comment regarding the name of the generic attribute.

Please, also resolve the conflicts and update this PR as some of the previous fields have changed (e.g. `server.socket.address`).

File: model/trace/database.yaml
Comment (arielvalentin) [reply model/trace/database.yaml]: `db.instance.id` works for me.

**Comment by arielvalentin** [comment]
@AlexanderWert I have rebased and resolved conflicts as well as renamed the attribute to `db.instance.id`.

File: docs/database/database-spans.md
Comment (AlexanderWert) [inline docs/database/database-spans.md:68]:
```suggestion
| `db.instance.id` | string | An identifier (address, unique name, or any other identifier) of the database instance that is executing queries or mutations on the current connection. This is useful in cases where the database is running in a clustered environment and the instrumentation is able to record the node executing the query. The client may obtain this value in databases like MySQL using queries like `select @@hostname`. | `mysql-e26b99z.example.com` | Recommended: If different from the `server.address` |
```

File: model/trace/database.yaml
Comment (AlexanderWert) [inline model/trace/database.yaml:253]:
```suggestion
          An identifier (address, unique name, or any other identifier) of the database instance that is executing queries or mutations on the current connection.
```

**Review by AlexanderWert — APPROVED** [review]
Nit: a small suggestion on the wording to make clear this field can contain any type of identifier (address, unique name, ids, etc.)

**Comment by arielvalentin** [comment]
@AlexanderWert thanks again for the review I have rebased and fixed the latest conflicts introduced by the registry refactoring.

Is there anything else that I need to do?

**Comment by AlexanderWert** [comment]
> @AlexanderWert thanks again for the review I have rebased and fixed the latest conflicts introduced by the registry refactoring.
> 
> Is there anything else that I need to do?

LGTM!

We just need more approvals.

File: docs/database/mysql.md
Comment (jsuereth) [reply docs/database/mysql.md]: I'd vote for `db.instance.address`

**Review by jsuereth — APPROVED** [review]

**Review by pyohannes — APPROVED** [review]

File: docs/database/mysql.md
Comment (AlexanderWert) [reply docs/database/mysql.md]: @jsuereth Please see my comment here: https://github.com/open-telemetry/semantic-conventions/pull/345#discussion_r1393789543

**Pull request merged**
